src/serac/physics/solid_residual.hpp

```diff
@@ -0,0 +1,629 @@
+  /// @tparam T the type of data to store at each quadrature point
+  template <typename T>
+  using qdata_type = std::shared_ptr<QuadratureData<T>>;
+
+  /**
+   * @brief Construct a new SolidResidual object
+   *
+   * @param physics_name A name for the physics module instance
+   * @param mesh_tag The tag for the mesh in the StateManager to construct the physics module on
+   * @param shape_disp_space Shape displacement space
+   * @param test_space Test space
+   * @param parameter_fe_spaces Vector of parameters spaces
+   * @param parameter_names Vector of parameter names, must match size of parameter_fe_spaces
+   */
+  SolidResidual(std::string physics_name, std::string mesh_tag, const mfem::ParFiniteElementSpace& shape_disp_space,
+                const mfem::ParFiniteElementSpace& test_space,
+                std::vector<const mfem::ParFiniteElementSpace*> parameter_fe_spaces = {},
+                std::vector<std::string> parameter_names = {})
+      : Residual(physics_name), mesh_tag_(mesh_tag), mesh_(StateManager::mesh(mesh_tag_))
+  {
+    std::array<const mfem::ParFiniteElementSpace*, NUM_STATE_VARS + sizeof...(parameter_space)> trial_spaces;
+    trial_spaces[0] = &test_space;
+    trial_spaces[1] = &test_space;
+    trial_spaces[2] = &test_space;
+
+    SLIC_ERROR_ROOT_IF(
+        sizeof...(parameter_space) != parameter_names.size(),
+        axom::fmt::format("{} parameter spaces given in the template argument but {} parameter names were supplied.",
+                          sizeof...(parameter_space), parameter_names.size()));
+
+    if constexpr (sizeof...(parameter_space) > 0) {
+      for_constexpr<sizeof...(parameter_space)>(
+          [&](auto i) { trial_spaces[i + NUM_STATE_VARS] = parameter_fe_spaces[i]; });
+    }
+
+    residual_ = std::make_unique<ShapeAwareFunctional<shape_trial, test(trial, trial, trial, parameter_space...)>>(
+        &shape_disp_space, &test_space, trial_spaces);
+  }
+
+  /**
+   * @brief register a custom boundary integral calculation as part of the residual
+   *
+   * @tparam active_parameters a list of indices, describing which parameters to pass to the q-function
+   * @param qfunction a callable that returns the traction on a boundary surface
+   * @param optional_domain The domain over which the boundary integral is evaluated. If nothing is supplied the entire
+   * boundary is used.
+   * ~~~ {.cpp}
+   *
+   *  solid_mechanics.addCustomBoundaryIntegral(DependsOn<>{}, [](double t, auto position, auto displacement, auto
+   * acceleration, auto shape){ auto [X, dX_dxi] = position;
+   *
+   *     auto [u, du_dxi] = displacement;
+   *     auto f           = u * 3.0 (X[0] < 0.01);
+   *     return f;  // define a displacement-proportional traction at a given support
+   *  });
+   *
+   * ~~~
+   *
+   * @note This method must be called prior to completeSetup()
+   */
+  template <int... active_parameters, typename callable>
```

Comment:
Should additional terms in the residual be a part of `SolidResidual`?  Maybe each term should be its own class?